File: java/src/main/java/org/hyperledger/fabric/client/identity/Signers.java
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.hyperledger.fabric.client.identity;

import java.security.PrivateKey;
import java.security.interfaces.ECPrivateKey;

/**
 * Factory methods to create standard signing implementations.
 */
public final class Signers {
    private static final String ED25519_ALGORITHM = "Ed25519";

    /**
     * Create a new signer that uses the supplied private key for signing. The {@link Identities} class provides static
     * methods to create a {@code PrivateKey} object from PEM-format data.
     *
     * <p>Currently supported private key types are:</p>
     * <ul>
     *     <li>ECDSA.</li>
     *     <li>Ed25519.</li>
     * </ul>
     *
     * <p>Note that the Sign implementations have different expectations on the input data supplied to them.</p>
     *
     * <p>The ECDSA signers operate on a pre-computed message digest, and should be combined with an appropriate hash
     * algorithm. P-256 is typically used with a SHA-256 hash, and P-384 is typically used with a SHA-384 hash.</p>
     *
     * <p>The Ed25519 signer operates on the full message content, and should be combined with a
     * {@link org.hyperledger.fabric.client.Hash#NONE NONE} (or no-op) hash implementation to ensure the complete
     * message is passed to the signer.</p>
     * @param privateKey A private key.
     * @return A signer implementation.
     */
    public static Signer newPrivateKeySigner(final PrivateKey privateKey) {
        if (privateKey instanceof ECPrivateKey) {
            return new ECPrivateKeySigner((ECPrivateKey) privateKey);
        }

        if (ED25519_ALGORITHM.equals(privateKey.getAlgorithm())) {
            return new PrivateKeySigner(privateKey, ED25519_ALGORITHM);
        }

        throw new IllegalArgumentException("Unsupported private key type: " + privateKey.getClass().getTypeName()
                + " (" + privateKey.getAlgorithm() + ")");
    }

    // Private constructor to prevent instantiation
    private Signers() { }
}

File: web/src/identity/signers.ts
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { newECPrivateKeySigner } from './ecdsa';
import { Signer } from './signer';

/**
 * Create a new signing implementation that uses the supplied private key to sign messages.
 *
 * Currently supported private key types are:
 * - NIST P-256 elliptic curve.
 * - NIST P-384 elliptic curve.
 * - Ed25519.
 *
 * Note that the signer implementations have different expectations on the input data supplied to them.
 *
 * The P-256 and P-384 signers operate on a pre-computed message digest, and should be combined with an appropriate
 * hash algorithm. P-256 is typically used with a SHA-256 hash, and P-384 is typically used with a SHA-384 hash.
 *
 * The Ed25519 signer operates on the full message content, and should be combined with a {@link hash.none | none} (or
 * no-op) hash implementation to ensure the complete message is passed to the signer.
 *
 * @param key - A private key.
 * @returns A signing implementation.
 */
export async function newPrivateKeySigner(key: CryptoKey): Promise<Signer> {
    if (key.type !== 'private') {
        throw new Error(`Invalid key type: ${key.type}`);
    }

    switch (key.algorithm.name) {
        case 'ECDSA':
            return newECPrivateKeySigner(key);
        case 'Ed25519':
            return async (message: Uint8Array) => {
                const signature = new Uint8Array(await crypto.subtle.sign('Ed25519', key, message));
                return Promise.resolve(signature);
            };
        default:
            throw new Error(`Unsupported private key type: ${String(key.algorithm.name)}`);
    }
}
